### DISCUSSION TITLE
User passwords encryption

### DISCUSSION THREAD

**Discussion by alanzaf** [Libki/libki-server#339] [Q&A]
Hi,
i need to create some hundreds of users from data i already have in csv format and had a look at users table and how data is stored. It looks like passwords are encrypted or encoded in some way, not stored in clear. I need to know how to encode my passwords for storing them in the table.
By the way, i guess i can simply load this users table with my users info, right? Anything else to do?
Thanks in advance,
Kind regards

**Comment by kylemhall** [comment]
The default digest algorithm is MD5 Base 64. Yes, I believe you are correct about only needing to load data into the users table!

**Reply by alanzaf** [reply]
Thanks Kyle, for your reply.
But it doesn't seem to work. I have set "one_password" as the password for a test user, and it's been encoded as "CK+xp0mg6ziZ5utnPAsoQQ"
Now, if i md5sum and encode it base64, i get a different string.

root@b797e98621d5:/app# echo "one_password" | md5sum | base64
NzJhNGJiOTg1MjRhMTcyYWFjMGExYjA0ZWQ5MmI5OTEgIC0K

Maybe i'm doing something wrong...

**Comment by kylemhall** [comment]
The password "test" comes out as CY9rzUYh03PK3k6DJie09g

`echo -n test | openssl dgst -md5 -binary | openssl enc -base64` outputs
CY9rzUYh03PK3k6DJie09g==

so does `(echo 0:; echo -n test | md5sum) | xxd -rp -l 16 | base64`

The equals signs at the end are padding you'll need to remove with a regular expression or something.

There is also a script at script/administration/import_users.pl that may be of use to you. It currently only supports username and password iirc but could be extended.

**Comment by alanzaf** [comment]
I see... will encode passwords that way. Thank you very much Kyle for your kind reply. It's great help.
Yes, there's that script you mention but, as you say, only supports username and password. I'm afraid my programming skills are not enough to extend it :(

**Comment by kylemhall** [comment]
If you just need additional columns from the users table, I'd be happy to update that script to include them!

**Reply by alanzaf** [reply]
Thank you Kyle, you're so kind, but it won't be necessary right now, as i already encoded my users passwords and loaded the users table manually, so to say. Thanks for the tips on encoding passwords.
